### PR TITLE
lifecycle: Prevent some configurations

### PR DESCRIPTION
**Pull request by AymericDu**
##### SUMMARY

- Prevent some configurations
  * `Date` and `Days` must not be mixed in the same Rule
  * `Date` in the Expiration action must be later than `Date` in the Transition action
  * `Days` in the Expiration action must be greater than `Days` in the Transition action
  * `NoncurrentDays` and `NoncurrentCount` must not be mixed  in the same Rule
  * `NoncurrentDays` in the NoncurrentVersionExpiration action must be greater than `NoncurrentDays` in the NoncurrentVersionTransition action
  * `NoncurrentCount` in the NoncurrentVersionExpiration action must be greater than `NoncurrentCount` in the NoncurrentVersionTransition action
- Allow multiple Transition and NoncurrentVersionTransition actions
- Apply Expiration and Transition actions only on the current version
- Format the Lifecycle configuration
- Add deleted flag when applying Expiration action with versioning
- Apply the rules in the same order as the configuration

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

- Lifecycle

##### SDS VERSION

```
openio 4.3.1.dev9
```

### CODE REVIEW

File: oio/container/lifecycle.py [hunk 1]
```diff
@@ -141,44 +187,168 @@ def execute(self, **kwargs):
                         "Failed to apply lifecycle rules on %s/%s/%s: %s",
                         self.account, self.container, obj_meta['name'], exc)
                 yield obj_meta, "n/a", "n/a", exc
+            if self.cache is not None:
+                self.cache.save_object(obj_meta, **kwargs)
+        self.cache = None
+
+    def is_current_version(self, obj_meta, **kwargs):
```
Comment (fvennetier) [inline oio/container/lifecycle.py]: No docstring?

File: oio/container/lifecycle.py [hunk 2]
```diff
@@ -141,44 +187,168 @@ def execute(self, **kwargs):
...
+    def from_element(cls, rule_elt, **kwargs):
         """
         Load the rule from an XML element.
 
         :type rule_elt: `lxml.etree.Element`
         """
-        filter_elt = rule_elt.find('Filter')
-        if filter_elt is None:
+        try:
+            id_ = rule_elt.findall('ID')[-1].text
```
Comment (fvennetier) [inline oio/container/lifecycle.py]: Is it necessary to call `findall`? Why not just `find`?

File: oio/container/lifecycle.py [hunk 3]
```diff
@@ -25,14 +27,45 @@
 from oio.common.utils import cid_from_name, depaginate
 
 
-LIFECYCLE_PROPERTY_KEY = "X-Container-Sysmeta-Swift3-Lifecycle"
-TAGGING_KEY = "x-object-sysmeta-swift3-tagging"
+STATUS_EXPECTED = ['enabled', 'disabled']
```
Comment (fvennetier) [inline oio/container/lifecycle.py]: I think you mean `ALLOWED_STATUSES`.

File: oio/cli/lifecycle/lifecycle.py [hunk 4]
```diff
@@ -107,8 +113,14 @@ def get_parser(self, prog_name):
...
-        self.app.stdout.write(md['properties'].get(LIFECYCLE_PROPERTY_KEY, ''))
+        lc = ContainerLifecycle(self.app.client_manager.storage,
+                                self.app.client_manager.account,
+                                parsed_args.container,
+                                self.log)
+        xml = lc.get()
+        if xml is None:
+            raise Exception(
+                "No lifecycle configuration for container %s in account %s" %
+                (parsed_args.container, self.app.client_manager.account))
```
[the remark this answers is not in the document]
Comment (AymericDu) [reply oio/cli/lifecycle/lifecycle.py]: I raise an exception to do like the AWS S3 CLI.
```
$ aws --profile amazon s3 mb "s3://openio.aymeric"
make_bucket: openio.aymeric
$ aws --profile amazon s3api get-bucket-lifecycle-configuration --bucket "openio.aymeric"

An error occurred (NoSuchLifecycleConfiguration) when calling the GetBucketLifecycleConfiguration operation: The lifecycle configuration does not exist
```

**Pull request merged**